### PR TITLE
[Android] Fix the issue of blank or frozen pages in shared engine scenarios

### PR DESCRIPTION
**Pull request by 0xZOne**
Consider this scenario: In an add-to-app context, where multiple Flutter activities share the same engine, a situation occurs. When navigating from FlutterActivity1 to FlutterActivity2, the Flutter view associated with FlutterActivity1 is detached from the engine. Then, the Flutter view of FlutterActivity2 is attached. Upon navigating back to FlutterActivity1, its Flutter view is re-attached to the shared engine.

The expected behavior is: When a Flutter view detaches from the shared engine, the associated surface should be released. When the Flutter view re-attaches, a new surface should be created.

After #47358, no new surface is created when the Flutter view is attached again. This results in the Flutter view having no underlying surface, which causes the page to appear blank or freeze without responding.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java [hunk 1]
```diff
@@ -215,8 +220,6 @@ public void detachFromRenderer() {
         disconnectSurfaceFromRenderer();
       }
 
-      pause();
```
Comment (johnmccutchan) [inline shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java:218]: Why are you removing the calls to pause() from detach?

**Review by johnmccutchan — COMMENTED** [review]
> The expected behavior is: When a Flutter view detaches from the shared engine, the associated surface should be released. When the Flutter view re-attaches, a new surface should be created.
> After https://github.com/flutter/engine/pull/47358, no new surface is created when the Flutter view is attached again. This results in the Flutter view having no underlying surface, which causes the page to appear blank or freeze without responding.

Given your description, I would expect you to fix the bug that no new surface is created when the view is attached again. It looks like you've just removed calls to pause() which isn't addressing the root problem.

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java (hunk 1 above)
Comment (0xZOne) [reply shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java:218]: The call to pause() set `isPaused` to `true`. 
```java
  public void pause() {
    if (flutterRenderer == null) {
      Log.w(TAG, "pause() invoked when no FlutterRenderer was attached.");
      return;
    }
    isPaused = true;
  }
```

Having `isPaused` set to `true` prevents the `startRenderingToSurface` function from creating a new surface.


```java
  // FlutterRenderer and getSurfaceTexture() must both be non-null.
  private void connectSurfaceToRenderer() {

    ...

    flutterRenderer.startRenderingToSurface(getHolder().getSurface(), isPaused);
  }
```

```java
  public void startRenderingToSurface(@NonNull Surface surface, boolean onlySwap) {
    ...

    this.surface = surface;

    if (onlySwap) {
      // In the swap case we are just swapping the surface that we render to.
      flutterJNI.onSurfaceWindowChanged(surface);
    } else {
      // In the non-swap case we are creating a new surface to render to.
      flutterJNI.onSurfaceCreated(surface);
    }
  }
```

**Comment by 0xZOne** [comment]
> > The expected behavior is: When a Flutter view detaches from the shared engine, the associated surface should be released. When the Flutter view re-attaches, a new surface should be created.
> > After #47358, no new surface is created when the Flutter view is attached again. This results in the Flutter view having no underlying surface, which causes the page to appear blank or freeze without responding.
> 
> Given your description, I would expect you to fix the bug that no new surface is created when the view is attached again. It looks like you've just removed calls to pause() which isn't addressing the root problem.

It works fine after removing the call to pause() in the `detachFromRenderer` function. Calling `pause()` sets `isPaused` to true, which prevents `startRenderingToSurface` from creating a new surface.

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java (hunk 1 above)
Comment (johnmccutchan) [reply shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java:218]: What I'm asking you to do is fix the root bug and not just stop calling pause when it should be called.

**Review by johnmccutchan — CHANGES_REQUESTED** [review]

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java (hunk 1 above)
Comment (0xZOne) [reply shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java:218]: > What I'm asking you to do is fix the root bug and not just stop calling pause when it should be called.

I believe that I am **INDEED** addressing the root bug.

As stated in the comments of the `detachFromRenderer` function, if the `detachFromRenderer` function is called, it indicates that the `RenderSurface` is no longer needed and its underlying surface will also be detached. 

In my view, calling `pause()` within `detachFromRenderer()` is superfluous and contributes nothing to the problem that https://github.com/flutter/engine/pull/47358 aims to solve.

I completely fail to understand why you insist on retaining `pause` in `detachFromRenderer`.


```java
  /**
   * Invoked by the owner of this {@code FlutterSurfaceView} when it no longer wants to render a
   * Flutter UI to this {@code FlutterSurfaceView}.
   *
   * <p>This method will cease any on-going rendering from Flutter to this {@code
   * FlutterSurfaceView}.
   */
  public void detachFromRenderer() {
    ...
  }
```

**Comment by johnmccutchan** [comment]
> > > The expected behavior is: When a Flutter view detaches from the shared engine, the associated surface should be released. When the Flutter view re-attaches, a new surface should be created.
> > > After #47358, no new surface is created when the Flutter view is attached again. This results in the Flutter view having no underlying surface, which causes the page to appear blank or freeze without responding.
> > 
> > 
> > Given your description, I would expect you to fix the bug that no new surface is created when the view is attached again. It looks like you've just removed calls to pause() which isn't addressing the root problem.
> 
> It works fine after removing the call to pause() in the `detachFromRenderer` function. Calling `pause()` sets `isPaused` to true, which prevents `startRenderingToSurface` from creating a new surface.

Why can't you just move the `isPaused = false;` so it executes before `connectSurfaceToRenderer` is called?

**Comment by 0xZOne** [comment]
> 
> Why can't you just move the `isPaused = false;` so it executes before `connectSurfaceToRenderer` is called?

On one hand, I believe the call to `pause()` in `detachFromRenderer` should be removed.

As stated in the comments of the `detachFromRenderer` function, if the `detachFromRenderer` function is called, it indicates that the `RenderSurface` is no longer needed and its underlying surface will also be detached. 

```java
  /**
   * Invoked by the owner of this {@code FlutterSurfaceView} when it no longer wants to render a
   * Flutter UI to this {@code FlutterSurfaceView}.
   *
   * <p>This method will cease any on-going rendering from Flutter to this {@code
   * FlutterSurfaceView}.
   */
  public void detachFromRenderer() {
    ...
  }
```

In my view, calling `pause()` in `detachFromRenderer()` is superfluous and contributes nothing to the problem that https://github.com/flutter/engine/pull/47358 aims to solve. Moreover, it results in the required new surface not being created. Therefore, the call to `pause()` in `detachFromRenderer` should be removed. 

On the other hand, if `isPaused = false;` is moved to execute before the call to `connectSurfaceToRenderer`, it would result in #47358 losing its effect.

**Comment by johnmccutchan** [comment]
Thanks for your patience and explanations. I believe you're correct and will land this PR soon

**Review by johnmccutchan — APPROVED** [review]

**Pull request merged**